src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc

```diff
@@ -453,5 +453,69 @@ std::string GcsAutoscalerStateManager::DebugString() const {
   return stream.str();
 }
 
+std::optional<
+    absl::flat_hash_map<ray::NodeID,
+                        std::vector<google::protobuf::Map<std::string, double>>>>
+GcsAutoscalerStateManager::GetPerNodeInfeasibleResourceRequests() const {
+  RAY_CHECK(thread_checker_.IsOnSameThread());
+
+  if (!autoscaling_state_.has_value()) {
+    return std::nullopt;
+  }
+
+  absl::flat_hash_map<ray::NodeID,
+                      std::vector<google::protobuf::Map<std::string, double>>>
+      per_node_infeasible_requests;
+
+  // obtain the infeasible requests from the autoscaler state
+  std::vector<google::protobuf::Map<std::string, double>>
+      autoscaler_infeasible_resource_shapes;
+  autoscaler_infeasible_resource_shapes.reserve(
+      autoscaling_state_.value().infeasible_resource_requests_size());
+  for (int i = 0; i < autoscaling_state_.value().infeasible_resource_requests_size();
+       i++) {
+    autoscaler_infeasible_resource_shapes.emplace_back(
+        autoscaling_state_.value().infeasible_resource_requests(i).resources_bundle());
+  }
+
+  // collect the infeasible requests per node
+  for (const auto &[node_id, time_resource_data_pair] : node_resource_info_) {
+    const auto &resource_load_by_shape =
+        time_resource_data_pair.second.resource_load_by_shape();
+
+    for (int i = 0; i < resource_load_by_shape.resource_demands_size(); i++) {
+      bool is_infeasible_shape = false;
+      auto &infeasible_resource_shape = autoscaler_infeasible_resource_shapes.at(0);
+      for (const auto &shape : autoscaler_infeasible_resource_shapes) {
+        if (resource_load_by_shape.resource_demands(i).num_infeasible_requests_queued() >
+                0 &&
+            MapEqual(shape, resource_load_by_shape.resource_demands(i).shape())) {
```

Comment:
nit: helps readability
```suggestion
        const auto &resource_demand = resource_load_by_shape.resource_demands(i);
        if (resource_demand.num_infeasible_requests_queued() > 0
            && MapEqual(shape, resource_demand.shape())) {
```